### PR TITLE
🐛 Save Temporal Workflow Id to the DB instead of the Workspace volume.

### PR DESCRIPTION
**Pull request by davinchia**
## What
A few days ago we removed the workflow volume from the Kubernetes deployment in order to simplify the set up. This also lets us do away with the `workspace` PVC.

In theory this isn't needed since the volume is mainly used for logs and our Kube deployment logs out to the Cloud storage. 

In the process of doing so, we realised the volume is used to store the temporal workflow id that is later used to cancel the workflow. 

This PR:
1) Adds a migration to add the `temporalWorkflowId` column to the `Attempts` table. Exposes various persistent methods for this.
2) Modify temporal to store the workflow id in this column. Modify cancellation to retrieve the workflow id from the table.

Things to call out:
1) This approach means the worker now requires access to the jobs DB. I think this is reasonable.
2) Some tests are disabled since we haven't really stabilised the Flyway + older file-base migrations yet. Follow up ticket has been created (#5857) and Liren is working on this.

## Recommended reading order
1. `Attempts.yaml` and `V0_29_15_001__Add_temporalWorkflowId_col_to_Attempts.java` to understand the schema changes.
2. `JobPersistence.java` and `DefaultJobPersistence.java` to understand how the new column is used.
3. `TemporalAttemptExecution.java` and `SchedulerHandler.java` to understand how cancellation is happening.

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector added to connector index like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

**Comment by davinchia** [comment]
@jrhizor heads up, turns out that this wasn't as straightforward as we thought. What do you think about the approach listed out?

I think it would also close #2564 at the same time. I looked into using Temporal's metadata store and it looks like this requires elastisearch, so using the default workflow id seems o-kay to me.

File: airbyte-scheduler/persistence/src/test/java/io/airbyte/scheduler/persistence/DefaultJobPersistenceTest.java
Comment (davinchia) [inline airbyte-scheduler/persistence/src/test/java/io/airbyte/scheduler/persistence/DefaultJobPersistenceTest.java:166]: @tuliren I'm realising that our current tests don't run Flyway migrations. I think we are going to have to add this to all the config tests too. Am I missing anything?

Will do in a follow up PR.

Comment (tuliren) [reply airbyte-scheduler/persistence/src/test/java/io/airbyte/scheduler/persistence/DefaultJobPersistenceTest.java:166]: I will look into it and create a follow up PR.

Comment (tuliren) [reply airbyte-scheduler/persistence/src/test/java/io/airbyte/scheduler/persistence/DefaultJobPersistenceTest.java:166]: Actually this is not unexpected. I know that I will miss updating some tests. So the plan is that we will deal with this when we write the first migration. I thought that I would be that person, but you actually did this for the first time. Sorry about the extra work.

**Comment by tuliren** [comment]
The failed unit test, `testJobsOfLastMigrationMatchSource`, is no longer relevant. I think you can delete it.

File: airbyte-scheduler/persistence/src/test/java/io/airbyte/scheduler/persistence/DefaultJobPersistenceTest.java
Comment (davinchia) [reply airbyte-scheduler/persistence/src/test/java/io/airbyte/scheduler/persistence/DefaultJobPersistenceTest.java:166]: no worries!

**Comment by davinchia** [comment]
@tuliren cool I'll delete it. Thanks!

We'll have to do the same for the other test method for the config database.

What's the best way for us to remember to clean this up? Is there an existing ticket or should I write one?

**Comment by tuliren** [comment]
I created a ticket: https://github.com/airbytehq/airbyte/issues/5857

File: airbyte-migration/src/test/java/io/airbyte/migrate/MigrationCurrentSchemaTest.java
Comment (davinchia) [inline airbyte-migration/src/test/java/io/airbyte/migrate/MigrationCurrentSchemaTest.java:67]: this will be tackled in https://github.com/airbytehq/airbyte/issues/5857

File: airbyte-server/src/test/java/io/airbyte/server/migration/RunMigrationTest.java
Comment (davinchia) [inline airbyte-server/src/test/java/io/airbyte/server/migration/RunMigrationTest.java:100]: @tuliren this also fails.

Reason:
1) StubDB is set up to use the latest schema from airbyte_db module.
2) Migrations fails since the migrations are from the old system and don't have the latest schemas.

Comment (davinchia) [reply airbyte-server/src/test/java/io/airbyte/server/migration/RunMigrationTest.java:100]: What if we 'freeze' the db schema in a different file so the migration tests pass for now?

Comment (davinchia) [reply airbyte-server/src/test/java/io/airbyte/server/migration/RunMigrationTest.java:100]: This is also making me wonder if we missed database archive migration.

I understand Flyway will handle database schema migration - is it also set up to handle migrating the database archive?

Comment (tuliren) [reply airbyte-server/src/test/java/io/airbyte/server/migration/RunMigrationTest.java:100]: No, Flyway cannot handle archive migration. It only runs in db. Once the table structure changes, the file data cannot be loaded into db.

I have a plan to handle this and created an issue to track it.

Comment (tuliren) [reply airbyte-server/src/test/java/io/airbyte/server/migration/RunMigrationTest.java:100]: > What if we 'freeze' the db schema in a different file so the migration tests pass? Think it's important to keep them until we get rid of this entirely.

Yes, we can pass in the last file based migration schema as the "latest" schema in this kind of unit test.

Comment (tuliren) [reply airbyte-server/src/test/java/io/airbyte/server/migration/RunMigrationTest.java:100]: Given all the questions you have raised, I think it is the best for the long term to completely deprecate the file based migrations and write Flyway replacement for them.

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalAttemptExecution.java
Comment (davinchia) [inline airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalAttemptExecution.java:114]: any other pieces of the code that also access the workflow directory do so via helpers that will create any missing directories, so this is safe to delete.

**Comment by davinchia** [comment]
@tuliren since this is the first time we are doing a Flyway migration, I ran test to confirm backwards compatibility with previous versions and everything seemed to work. I stood up Airbyte on 0.29.15-alpha, performed one sync, spun down the server, then bring it up at version `dev`. Saw the migration being applied and everything seemed to be working. Makes me think we are ok. 

Can you confirm this is alright? Slightly nervous since our migration tests are completely working.

**Review by jrhizor — APPROVED** [review]
Looks good. Only minor comments.

Are we only going to merge this after the migration tests are fixed? I'm a bit worried about having disabled migration tests over any releases.

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalAttemptExecution.java
Comment (cgardens) [reply airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalAttemptExecution.java:145]: @jrhizor I think because these jobs aren't saved in the database at all there would be no place to save the workflow id.

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalAttemptExecution.java
Comment (cgardens) [inline airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalAttemptExecution.java:146]: it feels like kinda big deal that we are accessing the jobs db from within temporal. not necessarily bad, but it is different from how we've been thinking about it. just want to make sure we do it consciously.

**Review by cgardens — APPROVED** [review]
nice!

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalAttemptExecution.java
Comment (davinchia) [reply airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalAttemptExecution.java:146]: You are right - it's a big change. See my above comment about accessing the workflow id from outside temporal/before the workflow is created.

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalAttemptExecution.java
Comment (davinchia) [reply airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalAttemptExecution.java:145]: What I couldn't figure out is how to retrieve the workflow id outside of the temporal workflow. I don't think that is possible since we aren't assigning a custom label/workflow id without standing up elastisearch. So we need the workflow to first be created (Temporal assigns an id) and then save it.

File: airbyte-workers/src/test/java/io/airbyte/workers/temporal/TemporalAttemptExecutionTest.java
Comment (davinchia) [reply airbyte-workers/src/test/java/io/airbyte/workers/temporal/TemporalAttemptExecutionTest.java:57]: I added a test in the acceptance tests.

I don't see why not. Will add one.

File: airbyte-server/src/test/java/io/airbyte/server/migration/RunMigrationTest.java
Comment (tuliren) [reply airbyte-server/src/test/java/io/airbyte/server/migration/RunMigrationTest.java:100]: Issue: https://github.com/airbytehq/airbyte/issues/5902

File: airbyte-migration/src/test/java/io/airbyte/migrate/MigrationCurrentSchemaTest.java
Comment (tuliren) [reply airbyte-migration/src/test/java/io/airbyte/migrate/MigrationCurrentSchemaTest.java:67]: I can work on this as soon as the Databricks connector is done. Should be at most a few hours work.

As long as no one is touching the file-based migration in the interim, we should be safe.

File: airbyte-scheduler/persistence/build.gradle
Comment (davinchia) [reply airbyte-scheduler/persistence/build.gradle:16]: aye this is confusing.

this dependency is basically needed everywhere we call `jobMigrator.migrate()`. in this case, it's L166 - 169 of the `DefaultJobPersistenceTest` java file.

**Comment by tuliren** [comment]
> @tuliren since this is the first time we are doing a Flyway migration, I ran test to confirm backwards compatibility with previous versions and everything seemed to work. I stood up Airbyte on 0.29.15-alpha, performed one sync, spun down the server, then bring it up at version `dev`. Saw the migration being applied and everything seemed to be working. Makes me think we are ok.
> 
> Can you confirm this is alright? Slightly nervous since our migration tests are completely working.

Yes, I think this is alright. Also I will add an acceptance test template for Flyway migration. Issue here: https://github.com/airbytehq/airbyte/issues/5903 Will work on this when I copy the job attempt state to the configs database. This should make new migration easier to test.

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/jobs/migrations/V0_29_15_001__Add_temporalWorkflowId_col_to_Attempts.java
Comment (tuliren) [reply airbyte-db/lib/src/main/java/io/airbyte/db/instance/jobs/migrations/V0_29_15_001__Add_temporalWorkflowId_col_to_Attempts.java:41]: > `SQLDataType.VARCHAR.nullable(false)`

This is not backward compatible. Ideally we should add nullable columns first.

File: airbyte-scheduler/persistence/src/main/java/io/airbyte/scheduler/persistence/DefaultJobPersistence.java
Comment (tuliren) [reply airbyte-scheduler/persistence/src/main/java/io/airbyte/scheduler/persistence/DefaultJobPersistence.java:307]: Agree that migrating to an optional field is safer.

File: airbyte-scheduler/persistence/build.gradle
Comment (tuliren) [reply airbyte-scheduler/persistence/build.gradle:16]: This is needed because some of the jobs persistence unit tests need to run the migration first. When we launch a database, it only has the initial schema.

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/jobs/migrations/V0_29_15_001__Add_temporalWorkflowId_col_to_Attempts.java
Comment (tuliren) [reply airbyte-db/lib/src/main/java/io/airbyte/db/instance/jobs/migrations/V0_29_15_001__Add_temporalWorkflowId_col_to_Attempts.java:41]: If we decide to use `VARCHAR`, do we know how long this string will be? Better set a limit for it if we have a rough idea about the format of the ID.

File: airbyte-db/lib/src/main/resources/jobs_database/schema_dump.txt
Comment (tuliren) [inline airbyte-db/lib/src/main/resources/jobs_database/schema_dump.txt:31]: Without limiting the length of the `VARCHAR`, the max length is really large and unnecessary.

File: airbyte-scheduler/persistence/src/main/java/io/airbyte/scheduler/persistence/DefaultJobPersistence.java
Comment (tuliren) [inline airbyte-scheduler/persistence/src/main/java/io/airbyte/scheduler/persistence/DefaultJobPersistence.java:303]: By the way, you can use the generate jOOQ code now, which should be easier to write than raw SQL query.

File: airbyte-migration/src/test/java/io/airbyte/migrate/MigrationCurrentSchemaTest.java
Comment (tuliren) [reply airbyte-migration/src/test/java/io/airbyte/migrate/MigrationCurrentSchemaTest.java:67]: Maybe add `@Ignored` to it for now?

Comment (davinchia) [reply airbyte-migration/src/test/java/io/airbyte/migrate/MigrationCurrentSchemaTest.java:67]: Will switch to adding @Disabled and leave a todo for you

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/jobs/migrations/V0_29_15_001__Add_temporalWorkflowId_col_to_Attempts.java
Comment (davinchia) [reply airbyte-db/lib/src/main/java/io/airbyte/db/instance/jobs/migrations/V0_29_15_001__Add_temporalWorkflowId_col_to_Attempts.java:41]: No real reason. I thought a default value is clearer. Nullable field is fine for me. Good point on the backward compatibility. Will switch to nullable.

> If we decide to use VARCHAR, do we know how long this string will be? Better set a limit for it if we have a rough idea about the format of the ID.

Not clear. I looked at the documentation and there are no explicit limits. The return type is string, which is why I've left it as a flexible varchar.

File: airbyte-scheduler/persistence/src/main/java/io/airbyte/scheduler/persistence/DefaultJobPersistence.java
Comment (davinchia) [reply airbyte-scheduler/persistence/src/main/java/io/airbyte/scheduler/persistence/DefaultJobPersistence.java:303]: haha I don't dare to do so since this PR already has a lot of moving pieces. I'll try it next time 😄

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/jobs/migrations/V0_29_15_001__Add_temporalWorkflowId_col_to_Attempts.java
Comment (tuliren) [reply airbyte-db/lib/src/main/java/io/airbyte/db/instance/jobs/migrations/V0_29_15_001__Add_temporalWorkflowId_col_to_Attempts.java:41]: If this field is UUID strings, and you don't want to declare it as UUID type, you can do `SQLDataType.VARCHAR(36)`.

File: airbyte-scheduler/persistence/src/main/java/io/airbyte/scheduler/persistence/DefaultJobPersistence.java
Comment (davinchia) [reply airbyte-scheduler/persistence/src/main/java/io/airbyte/scheduler/persistence/DefaultJobPersistence.java:307]: Charles, I actually want to defer to you or @jrhizor  on this.

I went with the RTE because the existing code doesn't return an optional. `getJob` unwraps the optional and throws an RTE if the job isn't present.

From my experience, there is a lag between Temporal picking up the job, and the attempt being created. Since the workflow updates the workflow id, it is possible for an attempt to exist without a workflow id.

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/jobs/migrations/V0_29_15_001__Add_temporalWorkflowId_col_to_Attempts.java
Comment (davinchia) [reply airbyte-db/lib/src/main/java/io/airbyte/db/instance/jobs/migrations/V0_29_15_001__Add_temporalWorkflowId_col_to_Attempts.java:41]: [Looks](https://stackoverflow.com/a/8295254) like there isn't real benefit to specifying a length beyond enforcing a constraint check. Since Temporal doesn't explicitly state this is a UUID, I'm leaning towards leaving this as a flexible varchar.

File: airbyte-workers/src/test/java/io/airbyte/workers/temporal/TemporalAttemptExecutionTest.java
Comment (davinchia) [reply airbyte-workers/src/test/java/io/airbyte/workers/temporal/TemporalAttemptExecutionTest.java:57]: Hum I tried to do this and realised since the cancellation handler polls temporal for to figure out if the job is cancelled, this is going to need a working temporal cluster to be a valuable test. Possible but don't think it's worth the effort given we have this in the acceptance tests.

File: airbyte-db/lib/src/main/resources/jobs_database/schema_dump.txt
Comment (davinchia) [reply airbyte-db/lib/src/main/resources/jobs_database/schema_dump.txt:31]: oh I didn't realise this was what was happening under the hood. I'm going to put a 256 limit then.

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/jobs/migrations/V0_29_15_001__Add_temporalWorkflowId_col_to_Attempts.java
Comment (davinchia) [reply airbyte-db/lib/src/main/java/io/airbyte/db/instance/jobs/migrations/V0_29_15_001__Add_temporalWorkflowId_col_to_Attempts.java]: I didn't realise this was being translated to an insane limit under the hood. I'm going to use 256 for now to give us some buffer.

**Comment by davinchia** [comment]
Big PR review so going to summarise everything.

@jrhizor with Liren's blessing I think it's okay to ship this without the migration tests since this is backwards compatible. As long as there are no additional migrations in the mean time we are ok. Liren is going to work on fixing this soon, so we should be ok.

Remaining bits for both Charles and Jared to weigh in:
- Should `getAttemptTemporalWorkflowId` return an optional?
- Do we feel okay with the workflow setting a field in the DB? Is there a better way of doing so?

Thank you for bearing with me on this!

**Comment by jrhizor** [comment]
If Liren's change is going to happen very short term, shouldn't it block this PR? If Liren's change isn't going to be merged very short term, then it doesn't seem safe to block any other change that might have a migration for that period.

File: airbyte-scheduler/persistence/src/main/java/io/airbyte/scheduler/persistence/DefaultJobPersistence.java
Comment (jrhizor) [reply airbyte-scheduler/persistence/src/main/java/io/airbyte/scheduler/persistence/DefaultJobPersistence.java:307]: Should we wait here for Temporal to pick up the job? Doing an `Optional` makes this just a "best effort" method of getting the workflow id, right?

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalAttemptExecution.java
Comment (davinchia) [reply airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalAttemptExecution.java:146]: Spoke offline with Jared. This is the next phase of work: https://github.com/airbytehq/airbyte/issues/5936

File: airbyte-scheduler/persistence/src/main/java/io/airbyte/scheduler/persistence/DefaultJobPersistence.java
Comment (davinchia) [reply airbyte-scheduler/persistence/src/main/java/io/airbyte/scheduler/persistence/DefaultJobPersistence.java:307]: I think not waiting is okay.

I thought about this and Optional does seem better, since we might retrieve this before this is created and throwing/catching an exception is awkward.

**Comment by davinchia** [comment]
Spoke offline to Jared - merging this is okay since it only blocks schema migrations in the config database for the json blob.

**Comment by davinchia** [comment]
The Kube test is passing locally. Going to merge.

**Pull request merged**
